Modules/xxlimited.c

```diff
@@ -152,153 +264,109 @@ xx_foo(PyObject *self, PyObject *args)
-    "xxlimited.Null",
-    0,               /* basicsize */
-    0,               /* itemsize */
-    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
-    Null_Type_slots
-};
 
-/* ---------- */
 
 /* List of functions defined in the module */
 
 static PyMethodDef xx_methods[] = {
-    {"roj",             xx_roj,         METH_VARARGS,
-        PyDoc_STR("roj(a,b) -> None")},
     {"foo",             xx_foo,         METH_VARARGS,
         xx_foo_doc},
-    {"new",             xx_new,         METH_VARARGS,
+    {"new",             xx_new,         METH_NOARGS,
         PyDoc_STR("new() -> new Xx object")},
     {NULL,              NULL}           /* sentinel */
 };
 
+
+/* The module itself */
+
 PyDoc_STRVAR(module_doc,
 "This is a template module just for instruction.");
 
 static int
 xx_modexec(PyObject *m)
 {
-    PyObject *o;
-
-    /* Due to cross platform compiler issues the slots must be filled
-     * here. It's required for portability to Windows without requiring
-     * C++. */
-    Null_Type_slots[0].pfunc = &PyBaseObject_Type;
-    Null_Type_slots[1].pfunc = PyType_GenericNew;
-    Str_Type_slots[0].pfunc = &PyUnicode_Type;
-
-    Xxo_Type = PyType_FromSpec(&Xxo_Type_spec);
-    if (Xxo_Type == NULL)
-        goto fail;
-
-    /* Add some symbolic constants to the module */
-    if (ErrorObject == NULL) {
-        ErrorObject = PyErr_NewException("xxlimited.error", NULL, NULL);
-        if (ErrorObject == NULL)
-            goto fail;
+    xx_state *state = PyModule_GetState(m);
+
+    state->Error_Type = PyErr_NewException("xxlimited.Error", NULL, NULL);
+    if (state->Error_Type == NULL) {
+        return -1;
     }
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "error", ErrorObject);
-
-    /* Add Xxo */
-    o = PyType_FromSpec(&Xxo_Type_spec);
-    if (o == NULL)
-        goto fail;
-    PyModule_AddObject(m, "Xxo", o);
-
-    /* Add Str */
-    o = PyType_FromSpec(&Str_Type_spec);
-    if (o == NULL)
-        goto fail;
-    PyModule_AddObject(m, "Str", o);
-
-    /* Add Null */
-    o = PyType_FromSpec(&Null_Type_spec);
-    if (o == NULL)
-        goto fail;
-    PyModule_AddObject(m, "Null", o);
+    PyModule_AddType(m, (PyTypeObject*)state->Error_Type);
+
+    state->Xxo_Type = PyType_FromModuleAndSpec(m, &Xxo_Type_spec, NULL);
+    if (state->Xxo_Type == NULL) {
+        return -1;
+    }
+    PyModule_AddType(m, (PyTypeObject*)state->Xxo_Type);
+
+    // Add the Str type. It is not needed from C code, so it is only
+    // added to the module dict.
+    // It does not inherit from "object" (PyObject_Type), but from "str"
+    // (PyUnincode_Type). The tuple of bases is constructed manually.
+    // (It would also possible to fill the Py_tp_base slot, but
+    // it would need to be done dynamically: in standard C, `&PyUnicode_Type`
+    // is not a constant expression .)
+    PyObject *bases = PyTuple_Pack(1, (PyObject *)&PyUnicode_Type);
+    if (bases == NULL) {
+        return -1;
+    }
```

Comment:
maybe the spec could be defined here to use the tp_base member?

Reply:
That doesn't quite work on Windows IIRC. But PyType_FromModuleAndSpec can take a single type nowadays, I'll switch to that.

Reply:
> That doesn't quite work on Windows IIRC

If you define the spec variable in the exec function, it should work with any compiler. The issue is only for "global" variables.